hibernate-core/src/main/java/org/hibernate/persister/entity/JoinedSubclassEntityPersister.java

```diff
@@ -1407,6 +1416,17 @@ public void pruneForSubclasses(TableGroup tableGroup, Map<String, EntityNameUse>
 		else {
 			tableReferenceJoins.removeIf( join -> !retainedTableReferences.contains( join.getJoinedTableReference() ) );
 		}
+
+		if ( needsTreatDiscriminator ) {
+			// We have to add the discriminator condition to the root table join to make sure
+			// we're only retrieving the correct subtype records from the shared joined table
+			final String discriminatorFragment = getPrunedTableExpression(
+					entityNameUses,
+					metamodel,
+					tableGroup.getPrimaryTableReference().getIdentificationVariable()
```

Comment:
I don't think this will always be the correct table reference. Isn't the discriminator only on the root table?

Reply:
Yes the discriminator condition should always use the root table alias.

I believed `tableGroup.getPrimaryTableReference()` always returned a reference to the root table for joined subclasses, with table reference joins to the subtype-specific tables. Am I wrong about that?

Reply:
FYI `AbstractEntityPersister#createRootTableGroup` when creating the table reference, when `needsDiscriminator()` is true (like in this case) always uses the root table name:
```
final TableReference rootTableReference = new NamedTableReference(
		needsDiscriminator() ? getRootTableName() : getTableName(),
		sqlAliasBase.generateNewAlias()
);
```

Reply:
Why not? If the discriminator condition is applied to the root table it should be fine.

Alternatively, there's always the subquery approach (though it might perform worse).